ucan-discussion.md

```diff
@@ -0,0 +1,146 @@
+
+Although it’s only unique across both services if the user uses the same key-pair for both services.
+
+A minor downside of tying the “bucket” ID to the user’s key-pair is that we’re creating a possibly confusing situation where a cryptographic identifier is being used for something which is really just a file path. Conceptually this might just be slightly odd to the user.
+
+There are two possible ways we could implement this:
+
+#### Option 2.1 - not stored in DB
+
+Use the user’s public key as the bucket name, but don’t store it in the DB.
+
+This is pretty much un-viable because:
+* If we don’t store it in the DB then we can’t make it unique to the user. So each time a user created a UCAN they could do so with a different public key, giving them (yet another) new bucket.
+* We have no way of knowing which bucket belongs to which user; we only know which bucket belongs to which key. This creates a situation with several distinct downsides:
+    - a. If the user loses their private key, they lose access to their bucket; we can’t implement any form of recovery.
+    - b. If the user’s private key is leaked, then they have no way of revoking it, so access to their bucket is permanently breached.
+    - c. The user could potentially create “backup” keys by creating UCANs with unlimited expiry times which delegate full privileges to another key. This helps mitigate problem (a), but makes problem (b) more likely.
+    - d. With no possibility of revocation or recovery, it conceptually breaks the idea that, like passwords, key-pairs can be rotated.
+
+#### Option 2.2 - stored in DB
+
+In this implementation (which is what’s currently implemented), when a user uploads the root DID the first time, we store the relationship between the user adding the root UCAN and the the DID.
+
+This solves all the problems of option 2.1, or at least leaves open the possibility of solving them by adding revocation/recovery functionality. (But still shares the overall downside of option 2 mentioned at the top.)
+
+
+### Option 3 - A UUID
+
+Another option is to generate a UUID for each user to use as their bucket name.
+
+The advantages of this are that it’s:
+1. Short
+2. Globally unique across services
+3. Removes the link between key-pairs and the bucket name, allowing key-pairs to be rotated.
+
+This UUID could be synced between web3.storage and nft.storage to allow each user to have a single bucket identifier which is global. But unlike with option 2, the user doesn’t sync this value for us.
+
+
+Discussion
+----------
+
+It seems that the one advantage of using the DID as the bucket name is essentially this: it gets the user to do the work of setting a common value between nft.storage and web3.storage for us. As far as I (Adam) can see (but I might be wrong!) there is no other advantage to this over the UUID solution, it just gives us that syncing for free. The trade-off is the blurring of parameters and potential confusion that it may cause.
```

Comment:
Does user gets to even see bucket id ? If no then surface for confusion is fairly limited and basically scoped to engineering team. In which case probably can be addressed by documenting the facts.

If user does get to see the bucket ID it might be worth pondering what such a confusion could lead to and if it would make sense to disguise that DID (e.g. under CID) to remove potential confusion. That way benefits of DID are kept and tradeoffs are averted.